{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Module 4 - Programming Assignment\n",
    "\n",
    "## General Directions\n",
    "\n",
    "1. You must follow the Programming Requirements outlined on Canvas.\n",
    "2. The Notebook should be cleanly and fully executed before submission.\n",
    "3. You should change the name of this file to be your JHED id. For example, `jsmith299.ipynb` although Canvas will change it to something else...\n",
    "4. You must follow the Programming Requirments for this course.\n",
    "\n",
    "<div style=\"background: lemonchiffon; margin:20px; padding: 20px;\">\n",
    "    <strong>Important</strong>\n",
    "    <p>\n",
    "        You should always read the entire assignment before beginning your work, so that you know in advance what the requested output will be and can plan your implementation accordingly.\n",
    "    </p>\n",
    "</div>\n",
    "\n",
    "<div style=\"color: white; background: #C83F49; margin:20px; padding: 20px;\">\n",
    "    <strong>Academic Integrity and Copyright</strong>\n",
    "    <p>You are not permitted to consult outside sources (Stackoverflow, YouTube, ChatGPT, etc.) or use \"code assistance\" (Co-Pilot, etc) to complete this assignment. By submitting this assignment for grading, you certify that the submission is 100% your own work, based on course materials, group interactions, instructor guidance. You agree to comply by the requirements set forth in the Syllabus, including, by reference, the JHU KSAS/WSE Graduate Academic Misconduct Policy.</p>\n",
    "    <p>Sharing this assignment either directly (e.g., email, github, homework site) or indirectly (e.g., ChatGPT, machine learning platform) is a violation of the copyright. Additionally, all such sharing is a violation the Graduate Academic Misconduct Policy (facilitating academic dishonesty is itself academic dishonesty), even after you graduate.</p>\n",
    "    <p>If you have questions or if you're unsure about the policy, ask via Canvas Inbox. In this case, being forgiven is <strong>not</strong> easier than getting permission and ignorance is not an excuse.</p>\n",
    "    <p>This assignment is copyright (&copy Johns Hopkins University &amp; Stephyn G. W. Butcher). All rights reserved.</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 683,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pprint import pprint"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Local Search - Genetic Algorithm\n",
    "\n",
    "There are some key ideas in the Genetic Algorithm.\n",
    "\n",
    "First, there is a problem of some kind that either *is* an optimization problem or the solution can be expressed in terms of an optimization problem.\n",
    "For example, if we wanted to minimize the function\n",
    "\n",
    "$$f(x) = \\sum (x_i - 0.5)^2$$\n",
    "\n",
    "where $n = 10$.\n",
    "This *is* an optimization problem. Normally, optimization problems are much, much harder.\n",
    "\n",
    "![Eggholder](http://www.sfu.ca/~ssurjano/egg.png)!\n",
    "\n",
    "The function we wish to optimize is often called the **objective function**.\n",
    "The objective function is closely related to the **fitness** function in the GA.\n",
    "If we have a **maximization** problem, then we can use the objective function directly as a fitness function.\n",
    "If we have a **minimization** problem, then we need to convert the objective function into a suitable fitness function, since fitness functions must always mean \"more is better\".\n",
    "\n",
    "Second, we need to *encode* candidate solutions using an \"alphabet\" analogous to G, A, T, C in DNA.\n",
    "This encoding can be quite abstract.\n",
    "You saw this in the Self Check.\n",
    "There a floating point number was encoded as bits, just as in a computer and a sophisticated decoding scheme was then required.\n",
    "\n",
    "Sometimes, the encoding need not be very complicated at all.\n",
    "For example, in the real-valued GA, discussed in the Lectures, we could represent 2.73 as....2.73.\n",
    "This is similarly true for a string matching problem.\n",
    "We *could* encode \"a\" as \"a\", 97, or '01100001'.\n",
    "And then \"hello\" would be:\n",
    "\n",
    "```\n",
    "[\"h\", \"e\", \"l\", \"l\", \"o\"]\n",
    "```\n",
    "\n",
    "or\n",
    "\n",
    "```\n",
    "[104, 101, 108, 108, 111]\n",
    "```\n",
    "\n",
    "or\n",
    "\n",
    "```\n",
    "0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1\n",
    "```\n",
    "\n",
    "In Genetics terminology, this is the **chromosome** of the individual. And if this individual had the **phenotype** \"h\" for the first character then they would have the **genotype** for \"h\" (either as \"h\", 104, or 01101000).\n",
    "\n",
    "To keep it straight, think **geno**type is **genes** and **pheno**type is **phenomenon**, the actual thing that the genes express.\n",
    "So while we might encode a number as 10110110 (genotype), the number itself, 182, is what goes into the fitness function.\n",
    "The environment operates on zebras, not the genes for stripes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## String Matching\n",
    "\n",
    "You are going to write a Genetic Algorithm that will solve the problem of matching a target string (at least at the start).\n",
    "Now, this is kind of silly because in order for this to work, you need to know the target string and if you know the target string, why are you trying to do it?\n",
    "Well, the problem is *pedagogical*.\n",
    "It's a fun way of visualizing the GA at work, because as the GA finds better and better candidates, they make more and more sense.\n",
    "\n",
    "Now, string matching is not *directly* an optimization problem so this falls under the general category of \"if we convert the problem into an optimization problem we can solve it with an optimization algorithm\" approach to problem solving.\n",
    "This happens all the time.\n",
    "We have a problem.\n",
    "We can't solve it.\n",
    "We convert it to a problem we *can* solve.\n",
    "In this case, we're using the GA to solve the optimization part.\n",
    "\n",
    "And all we need is some sort of measure of the difference between two strings.\n",
    "The only constraint for our objective function is that it must calculate the score based on element to element (character to character) comparisons with no global transformations of the candidate or target strings.\n",
    "That measure becomes our **objective function** and we can use it with the Genetic Algorithm.\n",
    "\n",
    "Since it is probably easier to come up with a score that measures how far apart the two strings are, we will end up with an objective function that represents a **minimization problem**.\n",
    "Because a fitness function must always be \"more is better\", we'll need to convert our objective function to a proper fitness function as well.\n",
    "\n",
    "And since this is a GA, we need a **genotype**.\n",
    "The genotype for this problem is a list of \"characters\" (individual letters aren't special in Python like they are in some other languages):\n",
    "\n",
    "```\n",
    "[\"h\", \"e\", \"l\", \"l\", \"o\"]\n",
    "```\n",
    "\n",
    "and the **phenotype** is the resulting string:\n",
    "\n",
    "```\n",
    "\"hello\"\n",
    "```\n",
    "\n",
    "In addition to the generic code and problem specific loss function, you'll need to pick parameters for the run.\n",
    "These parameters include:\n",
    "\n",
    "1. population size\n",
    "2. number of generations\n",
    "3. probability of crossover\n",
    "4. probability of mutation\n",
    "\n",
    "You will also need to pick a selection algorithm, either roulette wheel or tournament selection.\n",
    "In the later case, you will need a tournament size.\n",
    "This is all part of the problem.\n",
    "\n",
    "Every **ten** (10) generations, you should print out the fitness, genotype, and phenotype of the best individual in the population for the specific generation.\n",
    "The function should return the best individual *of the entire run*, using the same format.\n",
    "\n",
    "```\n",
    "{'genotype': ['x', 'd', 'j', 'l', 's'],\n",
    "  'fitness': 0.041666666666666664,\n",
    "  'phenotype': 'xdjls'},\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 684,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALPHABET = \"abcdefghijklmnopqrstuvwxyz \""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"genetic_algorithm\"></a>\n",
    "### genetic_algorithm\n",
    "\n",
    "**(**\n",
    "\n",
    "You can add as many parameters as you need to `genetic_algorithm`.\n",
    "The documentation should be filled out according to the programming assignment requirements for the course.\n",
    "You do not need to unit test this function.\n",
    "Remember that functions should \"only do one thing\", may not be more than 20 lines.\n",
    "Write helper functions with documentation and unit tests.\n",
    "\n",
    "A complex function and application like this has a lot of interesting potential:\n",
    "\n",
    "1. There are a lot of parameters We could reduce those by using a Dictionary.\n",
    "2. There are a lot of different possible behaviors, including problem specific behaviors. We could use higher order functions.\n",
    "\n",
    "Beyond these hints, I leave those decisions to you.\n",
    "\n",
    "\n",
    "*This is very Mission Impossible. After reading the directions in this Markdown cell, when the time is right, remove them  (everything between and including the parentheses) and replace with your documentation for `genetic_algorithm`! I have started you off.*\n",
    "\n",
    "**)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 685,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "from typing import List, Dict, Tuple, Callable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def encode_chrarr(pheno: str) -> List[str]:\n",
    "    return list(pheno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def decode_chararr(geno: List[str]) -> str:\n",
    "    return \"\".join(geno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 686,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_chromosome(chromosome_length: int, alphabet: str) -> List[str]:\n",
    "    return [random.choice(alphabet) for _ in range(chromosome_length)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_population(target: List[str], alphabet: str, population_size: int) -> List[List[str]]:\n",
    "    return [generate_chromosome(len(target), alphabet) for _ in range(population_size)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 687,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fitness_forward(chromosome: List[str], target: List[str]) -> float:\n",
    "    score = 0\n",
    "    for gene, target_gene in zip(chromosome, target):\n",
    "        score += abs(ord(gene) - ord(target_gene))\n",
    "    return 1 / (1 + score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 688,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fitness_reverse(chromosome: List[str], target: List[str]) -> float:\n",
    "    score = 0\n",
    "    chromosome_length = len(chromosome) - 1\n",
    "    for index in range(chromosome_length):\n",
    "        score += abs(ord(chromosome[chromosome_length - index]) - ord(target[index]))\n",
    "    return 1 / (1 + score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 689,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rot13(char: str) -> int:\n",
    "    return (ord(char) - ord(\"a\") + 13) % 26 + ord(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 690,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test\n",
    "assert [chr(rot13(c)) for c in \"abc\"] == [\"n\", \"o\", \"p\"]\n",
    "assert [chr(rot13(c)) for c in \"nop\"] == [\"a\", \"b\", \"c\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 691,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fitness_ceaser(chromosome: List[str], target: List[str]) -> float:\n",
    "    score = 0\n",
    "    for gene, target_gene in zip(chromosome, target):\n",
    "        score += abs(rot13(gene) - ord(target_gene))\n",
    "    return 1 / (1 + score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 692,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sort_population(population: List[List[str]], target: List[str], fitness: Callable) -> List[List[str]]:\n",
    "    return sorted(population, key=lambda chromosome: fitness(chromosome, target))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 693,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tournament(population: List[List[str]], target: List[str], sample_size: int, fitness: Callable):\n",
    "    sub_population = random.sample(population, k=sample_size)\n",
    "    sub_population = sort_population(sub_population, target, fitness)\n",
    "    return sub_population[-1], sub_population[-2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 694,
   "metadata": {},
   "outputs": [],
   "source": [
    "def recombine(chromosome1: List[str], chromosome2: List[str], cross_over_index: int, crossover_rand: float, crossover_rate: float) -> Tuple[List[str], List[str]]:\n",
    "    if crossover_rand >= crossover_rate:\n",
    "        return chromosome1, chromosome2\n",
    "    child1 = chromosome1[:cross_over_index] + chromosome2[cross_over_index:]\n",
    "    child2 = chromosome2[:cross_over_index] + chromosome1[cross_over_index:]  # crossover 2 list of strings\n",
    "    return child1, child2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 695,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate(chromosome: List[str], mutate_rand: float, mutate_rate: float) -> List[str]:\n",
    "    index = random.randint(0, len(chromosome) - 1)\n",
    "    if mutate_rand < mutate_rate:\n",
    "        chromosome[index] = random.choice(ALPHABET)\n",
    "    return chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def describe_chromosome(chromosome: List[str], target: List[str], fitness: Callable) -> Dict:\n",
    "    return {\"genotype\": chromosome, \"fitness\": fitness(chromosome, target), \"phenotype\": decode_chararr(chromosome)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_solution(population: List[List[str]], target: List[str], fitness: Callable) -> Dict: \n",
    "    best_solution = sort_population(population, target, fitness)[-1]\n",
    "    return describe_chromosome(best_solution, target, fitness)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def breed(parent1: List[str], parent2: List[str], crossover_rate: float, mutate_rate: float) -> Tuple[List[str], List[str]]:\n",
    "    child1, child2 = recombine(parent1, parent2, random.randint(0, len(parent1) - 1), random.uniform(0, 1), crossover_rate)\n",
    "    child1 = mutate(child1, random.uniform(0, 1), mutate_rate)\n",
    "    child2 = mutate(child2, random.uniform(0, 1), mutate_rate)\n",
    "    return child1, child2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genetic_algorithm(target_str: str, fitness: Callable, parameters: Dict, log_frequency: int = 10) -> Dict:\n",
    "    population_size, tournament_size, max_generation = parameters[\"population_size\"], parameters[\"tournament_size\"], parameters[\"max_generation\"] # ints\n",
    "    crossover_rate, mutate_rate, alphabet = parameters[\"crossover_rate\"], parameters[\"mutate_rate\"], parameters[\"alphabet\"]\n",
    "\n",
    "    target: List[str] = encode_chrarr(target_str)  # encode\n",
    "    population: List[List[str]] = generate_population(target, alphabet, population_size)\n",
    "\n",
    "    for generation in range(max_generation):\n",
    "        next_population = []\n",
    "        for _ in range(population_size // 2):\n",
    "            parent1, parent2 = tournament(population, target, tournament_size, fitness)\n",
    "\n",
    "            if parent1 == target:\n",
    "                return describe_chromosome(parent1, target, fitness)\n",
    "            if parent2 == target:\n",
    "                return describe_chromosome(parent2, target, fitness)\n",
    "\n",
    "            child1, child2 = breed(parent1, parent2, crossover_rate, mutate_rate)\n",
    "            next_population.extend([child1, child2])\n",
    "\n",
    "        population = next_population\n",
    "        if generation % log_frequency == 0:\n",
    "            print(f\"{generation:5} {get_solution(population, target, fitness)}\")\n",
    "\n",
    "    return get_solution(population, target, fitness)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1\n",
    "\n",
    "The target is the string \"this is so much fun\".\n",
    "The challenge, aside from implementing the basic algorithm, is deriving a fitness function based on \"b\" - \"p\" (for example).\n",
    "The fitness function should come up with a fitness score based on element to element comparisons between target v. phenotype."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 699,
   "metadata": {},
   "outputs": [],
   "source": [
    "target1 = \"this is so much fun\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 700,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up if you need it.\n",
    "params = {\"population_size\": 100, \"tournament_size\": 10, \"max_generation\": 1000, \"crossover_rate\": 0.9, \"mutate_rate\": 0.05, \"alphabet\": ALPHABET}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 701,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    0 {'genotype': ['x', 'h', 'e', 'x', ' ', 'm', 'y', 'b', 'x', 'd', 'e', 'q', 'x', 'f', 'p', ' ', 'p', 'r', 'r'], 'fitness': 0.004761904761904762, 'phenotype': 'xhex mybxdeqxfp prr'}\n",
      "   10 {'genotype': ['s', 'h', 'e', 'x', ' ', 'm', 'w', ' ', 's', 'p', 'e', 'j', 'v', 'b', 'c', ' ', 'l', 'r', 'r'], 'fitness': 0.008928571428571428, 'phenotype': 'shex mw spejvbc lrr'}\n",
      "   20 {'genotype': ['s', 'h', 'e', 'p', ' ', 'm', 'w', ' ', 's', 'p', 'c', 'j', 'v', 'd', 'c', ' ', 'c', 'r', 'm'], 'fitness': 0.00980392156862745, 'phenotype': 'shep mw spcjvdc crm'}\n",
      "   30 {'genotype': ['s', 'h', 'i', 'p', ' ', 'm', 'w', ' ', 's', 'p', 'c', 'n', 'v', 'd', 'c', ' ', 'f', 't', 'm'], 'fitness': 0.01098901098901099, 'phenotype': 'ship mw spcnvdc ftm'}\n",
      "   40 {'genotype': ['s', 'h', 'i', 'u', ' ', 'f', 'w', ' ', 's', 'p', ' ', 'n', 'v', 'd', 'c', ' ', 'f', 't', 'm'], 'fitness': 0.045454545454545456, 'phenotype': 'shiu fw sp nvdc ftm'}\n",
      "   50 {'genotype': ['s', 'h', 'i', 't', ' ', 'f', 'q', ' ', 's', 'p', ' ', 'n', 'v', 'd', 'c', ' ', 'f', 't', 'm'], 'fitness': 0.05263157894736842, 'phenotype': 'shit fq sp nvdc ftm'}\n",
      "   60 {'genotype': ['s', 'h', 'i', 't', ' ', 'f', 'q', ' ', 's', 'o', ' ', 'n', 'v', 'd', 'l', ' ', 'f', 't', 'm'], 'fitness': 0.058823529411764705, 'phenotype': 'shit fq so nvdl ftm'}\n",
      "   70 {'genotype': ['s', 'h', 'i', 't', ' ', 'k', 'q', ' ', 's', 'o', ' ', 'n', 'v', 'd', 'i', ' ', 'f', 't', 'm'], 'fitness': 0.07692307692307693, 'phenotype': 'shit kq so nvdi ftm'}\n",
      "   80 {'genotype': ['s', 'h', 'i', 't', ' ', 'k', 'q', ' ', 's', 'o', ' ', 'm', 'v', 'd', 'i', ' ', 'f', 't', 'm'], 'fitness': 0.08333333333333333, 'phenotype': 'shit kq so mvdi ftm'}\n",
      "   90 {'genotype': ['s', 'h', 'i', 't', ' ', 'j', 'q', ' ', 's', 'o', ' ', 'm', 'u', 'd', 'i', ' ', 'f', 't', 'm'], 'fitness': 0.1, 'phenotype': 'shit jq so mudi ftm'}\n",
      "  100 {'genotype': ['s', 'h', 'i', 't', ' ', 'j', 'q', ' ', 's', 'o', ' ', 'm', 'u', 'd', 'i', ' ', 'f', 't', 'm'], 'fitness': 0.1, 'phenotype': 'shit jq so mudi ftm'}\n",
      "  110 {'genotype': ['s', 'h', 'i', 't', ' ', 'j', 'q', ' ', 's', 'o', ' ', 'm', 'u', 'd', 'i', ' ', 'f', 't', 'm'], 'fitness': 0.1, 'phenotype': 'shit jq so mudi ftm'}\n",
      "  120 {'genotype': ['s', 'h', 'i', 't', ' ', 'j', 'q', ' ', 's', 'o', ' ', 'm', 'u', 'd', 'i', ' ', 'f', 't', 'm'], 'fitness': 0.1, 'phenotype': 'shit jq so mudi ftm'}\n",
      "  130 {'genotype': ['s', 'h', 'i', 't', ' ', 'j', 'q', ' ', 's', 'o', ' ', 'm', 'u', 'd', 'i', ' ', 'f', 't', 'm'], 'fitness': 0.1, 'phenotype': 'shit jq so mudi ftm'}\n",
      "  140 {'genotype': ['s', 'h', 'i', 't', ' ', 'j', 'q', ' ', 's', 'o', ' ', 'm', 'u', 'd', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.125, 'phenotype': 'shit jq so mudh ftn'}\n",
      "  150 {'genotype': ['u', 'h', 'i', 't', ' ', 'j', 's', ' ', 's', 'o', ' ', 'm', 'u', 'd', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.16666666666666666, 'phenotype': 'uhit js so mudh ftn'}\n",
      "  160 {'genotype': ['u', 'h', 'i', 't', ' ', 'j', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.2, 'phenotype': 'uhit js so much ftn'}\n",
      "  170 {'genotype': ['u', 'h', 'i', 't', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.25, 'phenotype': 'uhit is so much ftn'}\n",
      "  180 {'genotype': ['u', 'h', 'i', 't', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.25, 'phenotype': 'uhit is so much ftn'}\n",
      "  190 {'genotype': ['u', 'h', 'i', 't', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.25, 'phenotype': 'uhit is so much ftn'}\n",
      "  200 {'genotype': ['u', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'uhis is so much ftn'}\n",
      "  210 {'genotype': ['u', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'uhis is so much ftn'}\n",
      "  220 {'genotype': ['u', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'uhis is so much ftn'}\n",
      "  230 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  240 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  250 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  260 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  270 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  280 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  290 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  300 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  310 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  320 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  330 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  340 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  350 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  360 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  370 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n",
      "  380 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 't', 'n'], 'fitness': 0.5, 'phenotype': 'this is so much ftn'}\n"
     ]
    }
   ],
   "source": [
    "result1 = genetic_algorithm(target1, fitness_forward, params)  # do what you need to do for your implementation but don't change the lines above or below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 702,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'fitness': 1.0,\n",
      " 'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u',\n",
      "              'c', 'h', ' ', 'f', 'u', 'n'],\n",
      " 'phenotype': 'this is so much fun'}\n"
     ]
    }
   ],
   "source": [
    "pprint(result1, compact=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2\n",
    "\n",
    "You should have working code now.\n",
    "The goal here is to think a bit more about fitness functions.\n",
    "The target string is now, 'nuf hcum os si siht'.\n",
    "This is obviously target #1 but reversed.\n",
    "If we just wanted to match the string, this would be trivial.\n",
    "Instead, this problem, we want to \"decode\" the string so that the best individual displays the target forwards.\n",
    "In order to do this, you'll need to come up with a fitness function that measures how successful candidates are towards this goal.\n",
    "The constraint is that you may not perform any global operations on the target or individuals.\n",
    "Your fitness function must still compare a single gene against a single gene.\n",
    "Your solution will likely not be Pythonic but use indexing.\n",
    "That's ok.\n",
    "<div style=\"background: lemonchiffon; margin:20px; padding: 20px;\">\n",
    "    <strong>Important</strong>\n",
    "    <p>\n",
    "        You may not reverse an entire string (either target or candidate) at any time.\n",
    "        Everything must be a computation of one gene against one gene (one letter against one letter).\n",
    "        Failure to follow these directions will result in 0 points for the problem.\n",
    "    </p>\n",
    "</div>\n",
    "\n",
    "The best individual in the population is the one who expresses this string *forwards*.\n",
    "\n",
    "\"this is so much fun\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 703,
   "metadata": {},
   "outputs": [],
   "source": [
    "target2 = \"nuf hcum os si siht\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 704,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up if you need it.\n",
    "params = {\"population_size\": 100, \"tournament_size\": 10, \"max_generation\": 1000, \"crossover_rate\": 0.9, \"mutate_rate\": 0.05, \"alphabet\": ALPHABET}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 705,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    0 {'genotype': ['g', 'j', 'k', 'o', ' ', 'm', 'm', 'b', 'z', 'n', 'k', 'v', 'f', 'g', 'x', ' ', 'f', 'h', 'k'], 'fitness': 0.0043859649122807015, 'phenotype': 'gjko mmbznkvfgx fhk'}\n",
      "   10 {'genotype': ['g', 'j', 'k', 'u', ' ', 'h', 't', ' ', 'w', 'r', ' ', 'm', 'u', 'h', 'b', ' ', 'a', 'v', 'o'], 'fitness': 0.029411764705882353, 'phenotype': 'gjku ht wr muhb avo'}\n",
      "   20 {'genotype': ['g', 'h', 'k', 'u', ' ', 'h', 't', ' ', 'u', 'o', ' ', 'm', 'u', 'h', 'b', ' ', 'e', 'v', 'o'], 'fitness': 0.043478260869565216, 'phenotype': 'ghku ht uo muhb evo'}\n",
      "   30 {'genotype': ['g', 'h', 'k', 'u', ' ', 'h', 't', ' ', 'u', 'o', ' ', 'm', 'u', 'a', 'c', ' ', 'e', 'v', 'm'], 'fitness': 0.05263157894736842, 'phenotype': 'ghku ht uo muac evm'}\n",
      "   40 {'genotype': ['g', 'h', 'i', 'u', ' ', 'h', 't', ' ', 'u', 'o', ' ', 'm', 'u', 'a', 'c', ' ', 'f', 'v', 'm'], 'fitness': 0.0625, 'phenotype': 'ghiu ht uo muac fvm'}\n",
      "   50 {'genotype': ['g', 'h', 'i', 'u', ' ', 'h', 't', ' ', 'u', 'o', ' ', 'm', 'u', 'd', 'c', ' ', 'f', 'v', 'm'], 'fitness': 0.06666666666666667, 'phenotype': 'ghiu ht uo mudc fvm'}\n",
      "   60 {'genotype': ['g', 'h', 'i', 's', ' ', 'h', 't', ' ', 'u', 'o', ' ', 'm', 'u', 'd', 'j', ' ', 'f', 'v', 'n'], 'fitness': 0.1111111111111111, 'phenotype': 'ghis ht uo mudj fvn'}\n",
      "   70 {'genotype': ['g', 'h', 'i', 's', ' ', 'h', 't', ' ', 't', 'o', ' ', 'm', 'u', 'c', 'j', ' ', 'f', 'v', 'n'], 'fitness': 0.14285714285714285, 'phenotype': 'ghis ht to mucj fvn'}\n",
      "   80 {'genotype': ['g', 'h', 'i', 's', ' ', 'h', 's', ' ', 't', 'o', ' ', 'm', 'u', 'c', 'j', ' ', 'f', 'v', 'n'], 'fitness': 0.16666666666666666, 'phenotype': 'ghis hs to mucj fvn'}\n",
      "   90 {'genotype': ['g', 'h', 'i', 's', ' ', 'h', 's', ' ', 't', 'o', ' ', 'm', 'u', 'c', 'j', ' ', 'f', 'v', 'n'], 'fitness': 0.16666666666666666, 'phenotype': 'ghis hs to mucj fvn'}\n",
      "  100 {'genotype': ['g', 'h', 'i', 's', ' ', 'h', 's', ' ', 't', 'o', ' ', 'm', 'u', 'c', 'j', ' ', 'f', 'v', 'n'], 'fitness': 0.16666666666666666, 'phenotype': 'ghis hs to mucj fvn'}\n",
      "  110 {'genotype': ['g', 'h', 'i', 's', ' ', 'h', 's', ' ', 't', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'v', 'n'], 'fitness': 0.25, 'phenotype': 'ghis hs to much fvn'}\n",
      "  120 {'genotype': ['g', 'h', 'i', 's', ' ', 'h', 's', ' ', 't', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'ghis hs to much fun'}\n",
      "  130 {'genotype': ['g', 'h', 'i', 's', ' ', 'h', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 0.5, 'phenotype': 'ghis hs so much fun'}\n",
      "  140 {'genotype': ['q', 'h', 'i', 's', ' ', 'h', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 0.5, 'phenotype': 'qhis hs so much fun'}\n",
      "  150 {'genotype': ['g', 'h', 'i', 's', ' ', 'h', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 0.5, 'phenotype': 'ghis hs so much fun'}\n",
      "  160 {'genotype': ['g', 'h', 'i', 's', ' ', 'h', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 0.5, 'phenotype': 'ghis hs so much fun'}\n",
      "  170 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  180 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  190 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  200 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  210 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  220 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  230 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  240 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  250 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  260 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  270 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  280 {'genotype': ['q', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'qhis is so much fun'}\n",
      "  290 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  300 {'genotype': ['a', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ahis is so much fun'}\n",
      "  310 {'genotype': ['k', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'khis is so much fun'}\n",
      "  320 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  330 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  340 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  350 {'genotype': ['u', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'uhis is so much fun'}\n",
      "  360 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  370 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  380 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  390 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  400 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  410 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  420 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  430 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  440 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  450 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  460 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  470 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  480 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  490 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  500 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  510 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  520 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  530 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  540 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  550 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  560 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  570 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  580 {'genotype': ['h', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'hhis is so much fun'}\n",
      "  590 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  600 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  610 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  620 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  630 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  640 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  650 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  660 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  670 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  680 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  690 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  700 {'genotype': ['v', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'vhis is so much fun'}\n",
      "  710 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'this is so much fun'}\n",
      "  720 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'this is so much fun'}\n",
      "  730 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'this is so much fun'}\n",
      "  740 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'this is so much fun'}\n",
      "  750 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'this is so much fun'}\n",
      "  760 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  770 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  780 {'genotype': ['v', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'vhis is so much fun'}\n",
      "  790 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  800 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  810 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  820 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  830 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  840 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  850 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  860 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  870 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  880 {'genotype': ['f', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'fhis is so much fun'}\n",
      "  890 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  900 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  910 {'genotype': ['e', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ehis is so much fun'}\n",
      "  920 {'genotype': ['e', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ehis is so much fun'}\n",
      "  930 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  940 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  950 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  960 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  970 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  980 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n",
      "  990 {'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 1.0, 'phenotype': 'ghis is so much fun'}\n"
     ]
    }
   ],
   "source": [
    "result2 = genetic_algorithm(target2, fitness_reverse, params)  # do what you need to do for your implementation but don't change the lines above or below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 706,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'fitness': 1.0,\n",
      " 'genotype': ['g', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u',\n",
      "              'c', 'h', ' ', 'f', 'u', 'n'],\n",
      " 'phenotype': 'ghis is so much fun'}\n"
     ]
    }
   ],
   "source": [
    "pprint(result2, compact=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3\n",
    "\n",
    "This is a variation on the theme of Problem 2.\n",
    "The Caeser Cypher replaces each letter of a string with the letter 13 characters down alphabet (rotating from \"z\" back to \"a\" as needed).\n",
    "This is also known as ROT13 (for \"rotate 13\").\n",
    "Latin did not have spaces (and the space is not continguous with the letters a-z) so we'll remove them from our alphabet.\n",
    "Again, the goal is to derive a fitness function that compares a single gene against a single gene, without global transformations.\n",
    "This fitness function assigns higher scores to individuals that correctly decode the target.\n",
    "\n",
    "<div style=\"background: lemonchiffon; margin:20px; padding: 20px;\">\n",
    "    <strong>Important</strong>\n",
    "    <p>\n",
    "        You may not apply ROT13 to an entire string (either target or candidate) at any time.\n",
    "        Everything must be a computation of one gene against one gene.\n",
    "        Failure to follow these directions will result in 0 points for the problem.\n",
    "    </p>\n",
    "</div>\n",
    "\n",
    "The best individual will express the target *decoded*.\n",
    "\n",
    "\"thisissomuchfun\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 707,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALPHABET3 = \"abcdefghijklmnopqrstuvwxyz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 708,
   "metadata": {},
   "outputs": [],
   "source": [
    "target3 = \"guvfvffbzhpusha\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 709,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up if you need it\n",
    "params = {\"population_size\": 100, \"tournament_size\": 10, \"max_generation\": 1000, \"crossover_rate\": 0.9, \"mutate_rate\": 0.05, \"alphabet\": ALPHABET3}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 710,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    0 {'genotype': ['x', 'j', 'q', 'v', 'k', 'q', 'o', 'z', 'g', 'w', 'j', 'c', 'b', 'w', 'v'], 'fitness': 0.012345679012345678, 'phenotype': 'xjqvkqozgwjcbwv'}\n",
      "   10 {'genotype': ['s', 'f', 'i', 'r', 'h', 'u', 's', 'n', 'k', 'u', 'c', 'g', 'd', 't', 't'], 'fitness': 0.047619047619047616, 'phenotype': 'sfirhusnkucgdtt'}\n",
      "   20 {'genotype': ['s', 'g', 'i', 's', 'h', 'u', 's', 'n', 'm', 'u', 'c', 'g', 'f', 't', 'q'], 'fitness': 0.08333333333333333, 'phenotype': 'sgishusnmucgftq'}\n",
      "   30 {'genotype': ['s', 'g', 'i', 's', 'h', 's', 's', 'n', 'm', 'u', 'c', 'g', 'f', 't', 'q'], 'fitness': 0.1, 'phenotype': 'sgishssnmucgftq'}\n",
      "   40 {'genotype': ['s', 'g', 'i', 's', 'h', 's', 's', 'n', 'm', 'u', 'c', 'g', 'f', 't', ' '], 'fitness': 0.14285714285714285, 'phenotype': 'sgishssnmucgft '}\n",
      "   50 {'genotype': ['s', 'g', 'i', 's', 'h', 's', 's', 'o', 'm', 'u', 'c', 'g', 'f', 't', ' '], 'fitness': 0.16666666666666666, 'phenotype': 'sgishssomucgft '}\n",
      "   60 {'genotype': ['t', 'g', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'g', 'f', 't', ' '], 'fitness': 0.25, 'phenotype': 'tgisissomucgft '}\n",
      "   70 {'genotype': ['t', 'g', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'g', 'f', 't', ' '], 'fitness': 0.25, 'phenotype': 'tgisissomucgft '}\n",
      "   80 {'genotype': ['t', 'g', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'g', 'f', 't', ' '], 'fitness': 0.25, 'phenotype': 'tgisissomucgft '}\n",
      "   90 {'genotype': ['t', 'g', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'g', 'f', 't', ' '], 'fitness': 0.25, 'phenotype': 'tgisissomucgft '}\n",
      "  100 {'genotype': ['t', 'g', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'g', 'f', 't', ' '], 'fitness': 0.25, 'phenotype': 'tgisissomucgft '}\n",
      "  110 {'genotype': ['t', 'g', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'g', 'f', 't', ' '], 'fitness': 0.25, 'phenotype': 'tgisissomucgft '}\n",
      "  120 {'genotype': ['t', 'g', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'g', 'f', 't', ' '], 'fitness': 0.25, 'phenotype': 'tgisissomucgft '}\n",
      "  130 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'g', 'f', 'u', ' '], 'fitness': 0.5, 'phenotype': 'thisissomucgfu '}\n",
      "  140 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'g', 'f', 'u', ' '], 'fitness': 0.5, 'phenotype': 'thisissomucgfu '}\n",
      "  150 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'g', 'f', 'u', ' '], 'fitness': 0.5, 'phenotype': 'thisissomucgfu '}\n",
      "  160 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  170 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  180 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  190 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  200 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  210 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  220 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  230 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  240 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  250 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  260 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  270 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  280 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  290 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  300 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  310 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  320 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  330 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  340 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  350 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  360 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  370 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  380 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  390 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  400 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  410 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  420 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  430 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  440 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  450 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  460 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  470 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  480 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  490 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  500 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  510 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  520 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  530 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  540 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  550 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  560 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  570 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  580 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  590 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  600 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  610 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  620 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  630 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  640 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  650 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  660 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  670 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  680 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  690 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  700 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  710 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  720 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  730 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  740 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  750 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  760 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  770 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  780 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  790 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  800 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  810 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  820 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  830 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  840 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  850 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  860 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  870 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  880 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  890 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  900 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  910 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  920 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  930 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  940 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  950 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  960 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  970 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  980 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n",
      "  990 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f', 'u', ' '], 'fitness': 1.0, 'phenotype': 'thisissomuchfu '}\n"
     ]
    }
   ],
   "source": [
    "result3 = genetic_algorithm(target3, fitness_ceaser, params)  # do what you need to do for your implementation but don't change the lines above or below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 711,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'fitness': 1.0,\n",
      " 'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f',\n",
      "              'u', ' '],\n",
      " 'phenotype': 'thisissomuchfu '}\n"
     ]
    }
   ],
   "source": [
    "pprint(result3, compact=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4\n",
    "\n",
    "There is no code for this problem.\n",
    "\n",
    "In Problem 3, we assumed we knew what the shift was in ROT-13.\n",
    "What if we didn't?\n",
    "Describe how you might solve that problem including a description of the solution encoding (chromosome and interpretation) and fitness function. Assume we can add spaces into the message."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**answer here**\n",
    "\n",
    "In order to capture the transformation needed to translate between the genotype and target, we can add another gene (or set of genes) to store transformation information. For string matching, we could store an extra gene that encode the amount to rotate the string. We would then adjust the fitness function to treat the first n genes as normal characters and the last one as a rotation amount. We can then use the machinery of the genetic algorithm to drive towards solutions that have the correct transformation. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Challenge\n",
    "\n",
    "**You do not need to do this problem and it won't be graded if you do. It's just here if you want to push your understanding.**\n",
    "\n",
    "The original GA used binary encodings for everything.\n",
    "We're basically using a Base 27 encoding.\n",
    "You could, however, write a version of the algorithm that uses an 8 bit encoding for each letter (ignore spaces as they're a bit of a bother).\n",
    "That is, a 4 letter candidate looks like this:\n",
    "\n",
    "```\n",
    "0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1\n",
    "```\n",
    "\n",
    "If you wrote your `genetic_algorithm` code general enough, with higher order functions, you should be able to implement it using bit strings instead of latin strings."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Before You Submit...\n",
    "\n",
    "1. Did you provide output exactly as requested?\n",
    "2. Did you re-execute the entire notebook? (\"Restart Kernel and Rull All Cells...\")\n",
    "3. If you did not complete the assignment or had difficulty please explain what gave you the most difficulty in the Markdown cell below.\n",
    "4. Did you change the name of the file to `jhed_id.ipynb`?\n",
    "\n",
    "Do not submit any other files."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "en605645",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "120px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": false,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
