{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Module 4 - Programming Assignment\n",
    "\n",
    "## General Directions\n",
    "\n",
    "1. You must follow the Programming Requirements outlined on Canvas.\n",
    "2. The Notebook should be cleanly and fully executed before submission.\n",
    "3. You should change the name of this file to be your JHED id. For example, `jsmith299.ipynb` although Canvas will change it to something else...\n",
    "4. You must follow the Programming Requirments for this course.\n",
    "\n",
    "<div style=\"background: lemonchiffon; margin:20px; padding: 20px;\">\n",
    "    <strong>Important</strong>\n",
    "    <p>\n",
    "        You should always read the entire assignment before beginning your work, so that you know in advance what the requested output will be and can plan your implementation accordingly.\n",
    "    </p>\n",
    "</div>\n",
    "\n",
    "<div style=\"color: white; background: #C83F49; margin:20px; padding: 20px;\">\n",
    "    <strong>Academic Integrity and Copyright</strong>\n",
    "    <p>You are not permitted to consult outside sources (Stackoverflow, YouTube, ChatGPT, etc.) or use \"code assistance\" (Co-Pilot, etc) to complete this assignment. By submitting this assignment for grading, you certify that the submission is 100% your own work, based on course materials, group interactions, instructor guidance. You agree to comply by the requirements set forth in the Syllabus, including, by reference, the JHU KSAS/WSE Graduate Academic Misconduct Policy.</p>\n",
    "    <p>Sharing this assignment either directly (e.g., email, github, homework site) or indirectly (e.g., ChatGPT, machine learning platform) is a violation of the copyright. Additionally, all such sharing is a violation the Graduate Academic Misconduct Policy (facilitating academic dishonesty is itself academic dishonesty), even after you graduate.</p>\n",
    "    <p>If you have questions or if you're unsure about the policy, ask via Canvas Inbox. In this case, being forgiven is <strong>not</strong> easier than getting permission and ignorance is not an excuse.</p>\n",
    "    <p>This assignment is copyright (&copy Johns Hopkins University &amp; Stephyn G. W. Butcher). All rights reserved.</p>\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1282,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pprint import pprint"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Local Search - Genetic Algorithm\n",
    "\n",
    "There are some key ideas in the Genetic Algorithm.\n",
    "\n",
    "First, there is a problem of some kind that either *is* an optimization problem or the solution can be expressed in terms of an optimization problem.\n",
    "For example, if we wanted to minimize the function\n",
    "\n",
    "$$f(x) = \\sum (x_i - 0.5)^2$$\n",
    "\n",
    "where $n = 10$.\n",
    "This *is* an optimization problem. Normally, optimization problems are much, much harder.\n",
    "\n",
    "![Eggholder](http://www.sfu.ca/~ssurjano/egg.png)!\n",
    "\n",
    "The function we wish to optimize is often called the **objective function**.\n",
    "The objective function is closely related to the **fitness** function in the GA.\n",
    "If we have a **maximization** problem, then we can use the objective function directly as a fitness function.\n",
    "If we have a **minimization** problem, then we need to convert the objective function into a suitable fitness function, since fitness functions must always mean \"more is better\".\n",
    "\n",
    "Second, we need to *encode* candidate solutions using an \"alphabet\" analogous to G, A, T, C in DNA.\n",
    "This encoding can be quite abstract.\n",
    "You saw this in the Self Check.\n",
    "There a floating point number was encoded as bits, just as in a computer and a sophisticated decoding scheme was then required.\n",
    "\n",
    "Sometimes, the encoding need not be very complicated at all.\n",
    "For example, in the real-valued GA, discussed in the Lectures, we could represent 2.73 as....2.73.\n",
    "This is similarly true for a string matching problem.\n",
    "We *could* encode \"a\" as \"a\", 97, or '01100001'.\n",
    "And then \"hello\" would be:\n",
    "\n",
    "```\n",
    "[\"h\", \"e\", \"l\", \"l\", \"o\"]\n",
    "```\n",
    "\n",
    "or\n",
    "\n",
    "```\n",
    "[104, 101, 108, 108, 111]\n",
    "```\n",
    "\n",
    "or\n",
    "\n",
    "```\n",
    "0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1\n",
    "```\n",
    "\n",
    "In Genetics terminology, this is the **chromosome** of the individual. And if this individual had the **phenotype** \"h\" for the first character then they would have the **genotype** for \"h\" (either as \"h\", 104, or 01101000).\n",
    "\n",
    "To keep it straight, think **geno**type is **genes** and **pheno**type is **phenomenon**, the actual thing that the genes express.\n",
    "So while we might encode a number as 10110110 (genotype), the number itself, 182, is what goes into the fitness function.\n",
    "The environment operates on zebras, not the genes for stripes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## String Matching\n",
    "\n",
    "You are going to write a Genetic Algorithm that will solve the problem of matching a target string (at least at the start).\n",
    "Now, this is kind of silly because in order for this to work, you need to know the target string and if you know the target string, why are you trying to do it?\n",
    "Well, the problem is *pedagogical*.\n",
    "It's a fun way of visualizing the GA at work, because as the GA finds better and better candidates, they make more and more sense.\n",
    "\n",
    "Now, string matching is not *directly* an optimization problem so this falls under the general category of \"if we convert the problem into an optimization problem we can solve it with an optimization algorithm\" approach to problem solving.\n",
    "This happens all the time.\n",
    "We have a problem.\n",
    "We can't solve it.\n",
    "We convert it to a problem we *can* solve.\n",
    "In this case, we're using the GA to solve the optimization part.\n",
    "\n",
    "And all we need is some sort of measure of the difference between two strings.\n",
    "The only constraint for our objective function is that it must calculate the score based on element to element (character to character) comparisons with no global transformations of the candidate or target strings.\n",
    "That measure becomes our **objective function** and we can use it with the Genetic Algorithm.\n",
    "\n",
    "Since it is probably easier to come up with a score that measures how far apart the two strings are, we will end up with an objective function that represents a **minimization problem**.\n",
    "Because a fitness function must always be \"more is better\", we'll need to convert our objective function to a proper fitness function as well.\n",
    "\n",
    "And since this is a GA, we need a **genotype**.\n",
    "The genotype for this problem is a list of \"characters\" (individual letters aren't special in Python like they are in some other languages):\n",
    "\n",
    "```\n",
    "[\"h\", \"e\", \"l\", \"l\", \"o\"]\n",
    "```\n",
    "\n",
    "and the **phenotype** is the resulting string:\n",
    "\n",
    "```\n",
    "\"hello\"\n",
    "```\n",
    "\n",
    "In addition to the generic code and problem specific loss function, you'll need to pick parameters for the run.\n",
    "These parameters include:\n",
    "\n",
    "1. population size\n",
    "2. number of generations\n",
    "3. probability of crossover\n",
    "4. probability of mutation\n",
    "\n",
    "You will also need to pick a selection algorithm, either roulette wheel or tournament selection.\n",
    "In the later case, you will need a tournament size.\n",
    "This is all part of the problem.\n",
    "\n",
    "Every **ten** (10) generations, you should print out the fitness, genotype, and phenotype of the best individual in the population for the specific generation.\n",
    "The function should return the best individual *of the entire run*, using the same format.\n",
    "\n",
    "```\n",
    "{'genotype': ['x', 'd', 'j', 'l', 's'],\n",
    "  'fitness': 0.041666666666666664,\n",
    "  'phenotype': 'xdjls'},\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1283,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALPHABET = \"abcdefghijklmnopqrstuvwxyz \""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"genetic_algorithm\"></a>\n",
    "### genetic_algorithm\n",
    "\n",
    "**(**\n",
    "\n",
    "You can add as many parameters as you need to `genetic_algorithm`.\n",
    "The documentation should be filled out according to the programming assignment requirements for the course.\n",
    "You do not need to unit test this function.\n",
    "Remember that functions should \"only do one thing\", may not be more than 20 lines.\n",
    "Write helper functions with documentation and unit tests.\n",
    "\n",
    "A complex function and application like this has a lot of interesting potential:\n",
    "\n",
    "1. There are a lot of parameters We could reduce those by using a Dictionary.\n",
    "2. There are a lot of different possible behaviors, including problem specific behaviors. We could use higher order functions.\n",
    "\n",
    "Beyond these hints, I leave those decisions to you.\n",
    "\n",
    "\n",
    "*This is very Mission Impossible. After reading the directions in this Markdown cell, when the time is right, remove them  (everything between and including the parentheses) and replace with your documentation for `genetic_algorithm`! I have started you off.*\n",
    "\n",
    "**)**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1284,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "from typing import List, Dict, Tuple, Callable"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"encode_chrarr\"></a>\n",
    "## encode_chrarr\n",
    "\n",
    "*`encode_chrarr` encode a string as a list of characters (strings of length 1)*  **Used by**: [genetic_algorithm](#genetic_algorithm)\n",
    "\n",
    "* **pheno** str: phenotype of the gene\n",
    "\n",
    "**returns** List[str] - genotype of the gene represented as a list of characters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1285,
   "metadata": {},
   "outputs": [],
   "source": [
    "def encode_chrarr(pheno: str) -> List[str]:\n",
    "    return list(pheno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1286,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tests\n",
    "assert encode_chrarr(\"hello world\") == [\"h\", \"e\", \"l\", \"l\", \"o\", \" \", \"w\", \"o\", \"r\", \"l\", \"d\"] # test 1 - keeps spaces\n",
    "assert encode_chrarr(\"\") == [] # test 2 - empty string\n",
    "assert encode_chrarr(\"üåãüêä‚õ∞\") == [\"üåã\", \"üêä\", \"‚õ∞\"] # test 3 - non-ascii chars"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"decode_chrarr\"></a>\n",
    "## decode_chrarr\n",
    "\n",
    "*`decode_chrarr` decode a gene (list of characters) into its phenotype (string)*  **Used by**: [describe_chromosome](#describe_chromosome)\n",
    "\n",
    "* **geno** List[str] - genotype of the gene represented as a list of characters \n",
    "\n",
    "**returns** List[str] - phenotype of gene represented as a string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1287,
   "metadata": {},
   "outputs": [],
   "source": [
    "def decode_chrarr(geno: List[str]) -> str:\n",
    "    return \"\".join(geno)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1288,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tests\n",
    "assert decode_chrarr([\"h\", \"e\", \"l\", \"l\", \"o\", \" \", \"w\", \"o\", \"r\", \"l\", \"d\"]) == \"hello world\" # test 1 - keeps spaces\n",
    "assert decode_chrarr([]) == \"\" # test 2 - empty string\n",
    "assert decode_chrarr([\"üåã\", \"üêä\", \"‚õ∞\"]) == \"üåãüêä‚õ∞\" # test 3 - non-ascii chars"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"generate_chromosome\"></a>\n",
    "## generate_chromosome\n",
    "\n",
    "*`generate_chromosome` generate a random chromosome by selecting characters from a provided alphabet*  **Used by**: [generate_population](#generate_population)\n",
    "\n",
    "* **chromosome_length** int - number of characters in chromosome\n",
    "* **alphabet** str - alphabet of characters in gene\n",
    "\n",
    "**returns** List[str] - genotype of the gene represented as a list of characters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1289,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_chromosome(chromosome_length: int, alphabet: str) -> List[str]:\n",
    "    return [random.choice(alphabet) for _ in range(chromosome_length)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1290,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(generate_chromosome(4, \"abc\")) == 4 # test 1 - len of chromosome\n",
    "assert all(chrom in \"abc\" for chrom in generate_chromosome(4, \"abc\")) # test 2 - chromosome contains specified chars\n",
    "assert len(generate_chromosome(-1, \"abc\")) == 0 # test 3 - non-positive length returns empty chromosome "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"generate_population\"></a>\n",
    "## generate_population\n",
    "\n",
    "*`generate_population` generate a list of chromosomes*  **Used by**: [genetic_algorithm](#genetic_algorithm)\n",
    "\n",
    "* **chromosome_length** int - number of characters in chromosome\n",
    "* **alphabet** str - alphabet of characters in gene\n",
    "* **population_size** int - number of chromosomes in the population\n",
    "\n",
    "**returns** List[List[str]] - list of chromosomes in a population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1291,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generate_population(chromosome_length: int, alphabet: str, population_size: int) -> List[List[str]]:\n",
    "    return [generate_chromosome(chromosome_length, alphabet) for _ in range(population_size)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1292,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tests\n",
    "assert len(generate_population(4, \"abc\", 2)) == 2 # test 1 - len of chromosome\n",
    "assert [chrom in \"abc\" for inner_list in generate_population(4, \"abc\", 4) for chrom in inner_list] # test 2 - chromosome contains specified chars\n",
    "assert len(generate_population(-1, \"abc\", -1)) == 0 # test 3 - non-positive length returns empty chromosome "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"fitness_forward\"></a>\n",
    "## fitness_forward\n",
    "\n",
    "*`fitness_forward` calculate a potential solution chromosomes fitness by comparing it to the target chromosome. The chromosomes are compared gene by gene, ie character-wise, by summing up the difference in ascii value between the solution and target gene. The ascii-difference is transformed by $\\frac{1}{1+f}$ so that the fitness function values increase as the solution chromosome becomes more similar to the target chromosome*  **Used by**: [genetic_algorithm](#genetic_algorithm)\n",
    "\n",
    "* **chromosome** List[str] - genotype of potential solution \n",
    "* **target** List[str] - genotype of target\n",
    "\n",
    "**returns** float - fitness score quantifying how similar the two chromosomes are "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1293,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fitness_forward(chromosome: List[str], target: List[str]) -> float:\n",
    "    score = 0\n",
    "    for gene, target_gene in zip(chromosome, target):\n",
    "        score += abs(ord(gene) - ord(target_gene))\n",
    "    return 1 / (1 + score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1294,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test\n",
    "assert fitness_forward([\"a\", \"b\", \"c\"], [\"x\", \"y\", \"z\"]) < fitness_forward([\"x\", \"s\", \"v\"], [\"x\", \"y\", \"z\"]) # test 1 - fitness score increases with similarity\n",
    "assert fitness_forward([\"a\", \"b\", \"c\"], [\"a\", \"b\", \"c\"]) == 1.0 # test 2 - fitness of perfect match is 1\n",
    "assert fitness_forward([\"#\"], [\"(\"]) # no error when non-alpha value is used "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"fitness_reverse\"></a>\n",
    "## fitness_reverse\n",
    "\n",
    "*`fitness_reverse` calculate a potential solution chromosomes fitness by comparing it to the target chromosome. The chromosomes are compared in reverse order, starting from the first gene in the solution and the last gene of the target. A heuristic is calculated by summing up the difference in ascii value between the solution and target genes. The ascii-difference is transformed by $\\frac{1}{1+f}$ so that the fitness function values increase as the solution chromosome becomes more similar to the target chromosome*  **Used by**: [genetic_algorithm](#genetic_algorithm)\n",
    "\n",
    "* **chromosome** List[str] - genotype of potential solution \n",
    "* **target** List[str] - genotype of target\n",
    "\n",
    "**returns** float - fitness score quantifying how similar the two chromosomes are "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1295,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fitness_reverse(chromosome: List[str], target: List[str]) -> float:\n",
    "    score = 0\n",
    "    chromosome_length = len(chromosome) \n",
    "    for index in range(chromosome_length):\n",
    "        score += abs(ord(chromosome[chromosome_length - index -1]) - ord(target[index]))\n",
    "    return 1 / (1 + score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1296,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert fitness_reverse([\"a\", \"b\", \"c\"], [\"x\", \"y\", \"z\"]) < fitness_reverse([\"x\", \"s\", \"v\"], [\"x\", \"y\", \"z\"]) # test 1 - fitness score increases with similarity\n",
    "assert fitness_reverse([\"c\", \"b\", \"a\"], [\"a\", \"b\", \"c\"]) == 1.0 # test 2 - fitness of perfect match is 1\n",
    "assert fitness_reverse([\"#\"], [\"(\"]) # no error when non-alpha value is used "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"rot13\"></a>\n",
    "## rot13\n",
    "\n",
    "*`rot13` replace a letter by the letter 13 characters down the alphabet*  **Used by**: [fitness_ceaser](#fitness_ceaser)\n",
    "\n",
    "* **char** str - character\n",
    "\n",
    "**returns** str - character 13 letters down from the input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1297,
   "metadata": {},
   "outputs": [],
   "source": [
    "def rot13(char: str) -> str:\n",
    "    return chr((ord(char) - ord(\"a\") + 13) % 26 + ord(\"a\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1298,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Test\n",
    "assert [rot13(c) for c in \"abc\"] == [\"n\", \"o\", \"p\"] # test 1 \n",
    "assert [rot13(c) for c in \"nop\"] == [\"a\", \"b\", \"c\"] # test 2 \n",
    "assert [rot13(c) for c in \"\"] == [] # test 3 empty string"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"fitness_caeser\"></a>\n",
    "## fitness_caeser\n",
    "\n",
    "*`fitness_caeser` calculate a potential solution chromosomes fitness by comparing it to the target chromosome. The chromosomes are compared gene (character) by gene. A heuristic is calculated by summing up the difference in ascii value between the solution and target genes, after transforming the solution gene using [rot13](#rot13). The ascii-difference is transformed by $\\frac{1}{1+f}$ so that the fitness function values increase as the solution chromosome becomes more similar to the target chromosome*  **Used by**: [genetic_algorithm](#genetic_algorithm)\n",
    "\n",
    "* **chromosome** List[str] - genotype of potential solution \n",
    "* **target** List[str] - genotype of target\n",
    "\n",
    "**returns** float - fitness score quantifying how similar the two chromosomes are "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1299,
   "metadata": {},
   "outputs": [],
   "source": [
    "def fitness_caeser(chromosome: List[str], target: List[str]) -> float:\n",
    "    score = 0\n",
    "    for gene, target_gene in zip(chromosome, target):\n",
    "        score += abs(ord(rot13(gene)) - ord(target_gene))\n",
    "    return 1 / (1 + score)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1300,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tests\n",
    "\n",
    "assert fitness_caeser([\"a\", \"b\", \"c\"], [\"x\", \"y\", \"z\"]) > fitness_caeser([\"x\", \"s\", \"v\"], [\"x\", \"y\", \"z\"]) # test 1 - fitness score increases with similarity - note that rot13 is happening\n",
    "assert fitness_caeser([\"a\", \"b\", \"c\"], [\"n\", \"o\", \"p\"]) == 1.0 # test 2 - fitness of perfect match is 1\n",
    "assert fitness_caeser([\"#\"], [\"(\"]) # no error when non-alpha value is used "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"sort_population\"></a>\n",
    "## sort_population\n",
    "\n",
    "*`sort_population` sort a population of chromosomes by fitness*  **Used by**: [tournament](#tournament), [get_solution](#get_solution)\n",
    "\n",
    "* **population** List[List[str]] - list of chromosomes of potential solution \n",
    "* **target** List[str] - genotype of target\n",
    "* **fitness** Callable - fitness function\n",
    "\n",
    "**returns** List[List[str]] - list of chromosomes of potential solution in sorted order by fitness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1301,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sort_population(population: List[List[str]], target: List[str], fitness: Callable) -> List[List[str]]:\n",
    "    return sorted(population, key=lambda chromosome: fitness(chromosome, target))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1302,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tests\n",
    "pop = [[\"a\", \"b\", \"c\"],[\"d\", \"e\", \"f\"]]\n",
    "target = [\"a\", \"b\", \"c\"]\n",
    "\n",
    "assert sort_population(pop, target, fitness_forward)[-1] == [\"a\", \"b\", \"c\"] \n",
    "assert sort_population(pop, target, fitness_forward)[0] == [\"d\", \"e\", \"f\"]\n",
    "assert sort_population([[]], [], fitness_forward) == [[]]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"tournament\"></a>\n",
    "## tournament\n",
    "\n",
    "*`tournament` use tournament selection to pick two chromosomes from a population. First, a sub-population is chosen at random, the sub-population is sorted by fitness and the two chromosomes with the highest fitness are returned*  **Used by** [genetic_algorithm](#genetic_algorithm): \n",
    "\n",
    "* **population** List[List[str]] - list of chromosomes of potential solution \n",
    "* **target** List[str] - genotype of target\n",
    "* **sample_size** int - size of tournament sub-population\n",
    "* **fitness** Callable - fitness function\n",
    "\n",
    "**returns** Tuple[List[str], List[str]] - two most fit chromosomes from randomly chosen sub-population"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1303,
   "metadata": {},
   "outputs": [],
   "source": [
    "def tournament(population: List[List[str]], target: List[str], sample_size: int, fitness: Callable) -> Tuple[List[str], List[str]]:\n",
    "    if sample_size < 2 or len(population) < 2:\n",
    "        return ([], [])\n",
    "    sub_population = random.sample(population, k=sample_size)\n",
    "    sub_population = sort_population(sub_population, target, fitness)\n",
    "    return sub_population[-1], sub_population[-2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1304,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tests\n",
    "pop = generate_population(3, ALPHABET, 100)\n",
    "target = [\"a\", \"b\", \"c\"]\n",
    "assert len(tournament(pop, target, 10, fitness_forward)) == 2\n",
    "\n",
    "p1, p2 = tournament(pop, target, 10, fitness_forward)\n",
    "assert fitness_forward(p2, target) < fitness_forward(p1, target) \n",
    "\n",
    "assert tournament([[]], [], 0, fitness_forward) ==  ([], [])\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"recombine\"></a>\n",
    "## recombine\n",
    "\n",
    "*`recombine` swap portions of two chromosomes with each other at a specific crossover_index, if crossover_rand is less than crossover_rate. For example, given two chromosomes `#####` and `$$$$$`, a crossover at index 2 would look like: `##$$$`, `$$###`.* **Used by** [breed](#breed): \n",
    "\n",
    "* **chromosomes1** List[str] - chromosomes of potential solution \n",
    "* **chromosomes2** List[str] - chromosomes of potential solution \n",
    "* **crossover_index** int - position at which to crossover the two chromosomes\n",
    "* **crossover_rand** float - random number to select if crossover event occurs\n",
    "* **crossover_rate** float - rate at which crossover event occurs \n",
    "\n",
    "**returns** Tuple[List[str], List[str]] - two possibly recombined chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1305,
   "metadata": {},
   "outputs": [],
   "source": [
    "def recombine(chromosome1: List[str], chromosome2: List[str], crossover_index: int, crossover_rand: float, crossover_rate: float) -> Tuple[List[str], List[str]]:\n",
    "    if crossover_rand >= crossover_rate:\n",
    "        return chromosome1, chromosome2\n",
    "    child1 = chromosome1[:crossover_index] + chromosome2[crossover_index:]\n",
    "    child2 = chromosome2[:crossover_index] + chromosome1[crossover_index:]  # crossover 2 list of strings\n",
    "    return child1, child2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1306,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tests\n",
    "p1 = [\"#\",\"#\",\"#\",\"#\",\"#\"]\n",
    "p2 = [\"$\",\"$\",\"$\",\"$\",\"$\"]\n",
    "assert recombine(p1, p2, 2, 0.4, 0.5) == ([\"#\",\"#\",\"$\",\"$\",\"$\"], [\"$\",\"$\",\"#\",\"#\",\"#\"])\n",
    "assert recombine(p1, p2, 2, 0.5, 0.5) == ([\"#\",\"#\",\"#\",\"#\",\"#\"], [\"$\",\"$\",\"$\",\"$\",\"$\"])\n",
    "assert recombine(p1, p2, 2, 0.6, 0.5) == ([\"#\",\"#\",\"#\",\"#\",\"#\"], [\"$\",\"$\",\"$\",\"$\",\"$\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"mutate\"></a>\n",
    "## mutate\n",
    "\n",
    "*`mutate` swap character in chromosome with a mutation character at zero-indexed position mutate_index, if mutate_rand is less than mutate_rate. For example, given a chromosomes `#####` a mutation at position 2 could be `##$##`.* **Used by** [breed](#breed):\n",
    "\n",
    "* **chromosomes** List[str] - chromosomes of potential solution \n",
    "* **mutate_index** int - random number to select where mutation event occurs (zero-indexed)\n",
    "* **mutation** str - character to mutate to\n",
    "* **mutate_rand** float - random number to select if mutation event occurs\n",
    "* **mutate_rate** float - random number to select if mutation event occurs\n",
    "\n",
    "**returns** List[str] - possibly mutated chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1307,
   "metadata": {},
   "outputs": [],
   "source": [
    "def mutate(chromosome: List[str], mutate_index: int, mutation: str, mutate_rand: float, mutate_rate: float) -> List[str]:\n",
    "    if mutate_rand < mutate_rate:\n",
    "        chromosome[mutate_index] = mutation\n",
    "    return chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1308,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tests\n",
    "\n",
    "assert mutate([\"#\",\"#\",\"#\",\"#\",\"#\"], 2, \"$\", 0.4, 0.5) == [\"#\",\"#\",\"$\",\"#\",\"#\"]\n",
    "assert mutate([\"#\",\"#\",\"#\",\"#\",\"#\"], 2, \"$\", 0.5, 0.5) == [\"#\",\"#\",\"#\",\"#\",\"#\"]\n",
    "assert mutate([\"#\",\"#\",\"#\",\"#\",\"#\"], 2, \"$\", 0.6, 0.5) == [\"#\",\"#\",\"#\",\"#\",\"#\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"breed\"></a>\n",
    "## breed\n",
    "\n",
    "*`breed` Create two modified children chromosomes from the parents by selectively performing recombination (based on crossover_rate) and mutation (based on mutate_rate).* **Used by** [genetic_algorithm](#genetic_algorithm):\n",
    "\n",
    "* **parent1** List[str] - chromosomes of potential solution \n",
    "* **parent2** List[str] - chromosomes of potential solution \n",
    "* **crossover_rate** float - rate at which crossover event occurs \n",
    "* **mutate_rate** float - random number to select if mutation event occurs\n",
    "\n",
    "**returns** Tuple[List[str], List[str]] - two possibly recombined or mutated chromosomes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1309,
   "metadata": {},
   "outputs": [],
   "source": [
    "def breed(parent1: List[str], parent2: List[str], crossover_rate: float, mutate_rate: float) -> Tuple[List[str], List[str]]:\n",
    "    child1, child2 = recombine(parent1, parent2, random.randint(0, len(parent1) - 1), random.uniform(0, 1), crossover_rate)\n",
    "    child1 = mutate(child1, random.randint(0, len(child1) - 1), random.choice(ALPHABET), random.uniform(0, 1), mutate_rate)\n",
    "    child2 = mutate(child2, random.randint(0, len(child2) - 1), random.choice(ALPHABET), random.uniform(0, 1), mutate_rate)\n",
    "    return child1, child2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1310,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tests\n",
    "p1 = [\"#\",\"#\",\"#\",\"#\",\"#\"]\n",
    "p2 = [\"$\",\"$\",\"$\",\"$\",\"$\"]\n",
    "\n",
    "assert len(breed(p1, p2, 0.4, 0.4)) == 2  # output is right size\n",
    "assert breed(p1, p2, 1.0, 0.1)[0] != p1 and breed(p1, p2, 1.0, 0.1)[1] != p2 # recombination happens\n",
    "assert breed(p1.copy(), p2.copy(), 0.1, 1.0)[0] != p1 and breed(p1.copy(), p2.copy(), 0.1, 1.0)[1] != p2 # mutation happens"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"describe_chromosome\"></a>\n",
    "## describe_chromosome\n",
    "\n",
    "*`describe_chromosome` return a dictionary containing the genotype, phenotype and fitness of a chromosome* **Uses** [decode_chrarr](#decode_chrarr) **Used by** [get_solution](#get_solution), [genetic_algorithm](#genetic_algorithm):\n",
    "\n",
    "* **chromosomes** List[str] - chromosomes of potential solution \n",
    "* **target** List[str] - genotype of target\n",
    "* **fitness** Callable - fitness function\n",
    "\n",
    "\n",
    "**returns** Dict - dictionary containing the genotype, phenotype and fitness of a chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1311,
   "metadata": {},
   "outputs": [],
   "source": [
    "def describe_chromosome(chromosome: List[str], target: List[str], fitness: Callable) -> Dict:\n",
    "    return {\"genotype\": chromosome, \"fitness\": fitness(chromosome, target), \"phenotype\": decode_chrarr(chromosome)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1312,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert describe_chromosome([\"a\", \"b\", \"c\"], [\"a\", \"b\", \"c\"], fitness_forward) == {\"genotype\": [\"a\", \"b\", \"c\"], \"fitness\": 1.0, \"phenotype\": \"abc\"}\n",
    "assert describe_chromosome([\"a\", \"b\", \"c\"], [\"x\", \"y\", \"z\"], fitness_forward) == {\"genotype\": [\"a\", \"b\", \"c\"], \"fitness\": fitness_forward([\"a\", \"b\", \"c\"], [\"x\", \"y\", \"z\"]), \"phenotype\": \"abc\"}\n",
    "assert describe_chromosome([], [], fitness_forward) == {\"genotype\": [], \"fitness\": 1.0, \"phenotype\": \"\"}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"get_solution\"></a>\n",
    "## get_solution\n",
    "\n",
    "*`get_solution` return a dictionary containing the genotype, phenotype and fitness of the most fit chromosome in a population.* **Uses** [sort_population](#sort_population) and [describe_chromosome](#describe_chromosome) **Used by** [genetic_algorithm](#genetic_algorithm):\n",
    "\n",
    "* **population** List[List[str]] - list of chromosomes of potential solution \n",
    "* **target** List[str] - genotype of target\n",
    "* **fitness** Callable - fitness function\n",
    "\n",
    "**returns** Dict - dictionary containing the genotype, phenotype and fitness of a chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1313,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_solution(population: List[List[str]], target: List[str], fitness: Callable) -> Dict: \n",
    "    best_solution = sort_population(population, target, fitness)[-1]\n",
    "    return describe_chromosome(best_solution, target, fitness)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1314,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Tests\n",
    "pop = [[\"a\", \"b\", \"c\"],[\"d\", \"e\", \"f\"]]\n",
    "assert get_solution(pop, [\"a\", \"b\", \"c\"], fitness_forward) == {\"genotype\": [\"a\", \"b\", \"c\"], \"fitness\": 1.0, \"phenotype\": \"abc\"} \n",
    "assert get_solution(pop, [\"g\", \"f\", \"g\"], fitness_forward) ==  {\"genotype\": [\"d\", \"e\", \"f\"], \"fitness\": fitness_forward([\"d\", \"e\", \"f\"], [\"g\", \"f\", \"g\"]), \"phenotype\": \"def\"}\n",
    "assert sort_population([[]], [], fitness_forward) == [[]]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"genetic_algorithm\"></a>\n",
    "## genetic_algorithm\n",
    "\n",
    "*`genetic_algorithm` The goal of this function it to demonstrate genetic algorithms by finding words that match a target word. At the core of genetic algorithms is representing solutions to the problem as gene. An array of values (normally binary) are used to represent each solution. By decoding the gene we obtain the phenotype or solution state. Mimicking biology, we assign each genotype a fitness, and apply evolutionary pressure to ensure only solutions with high fitness selected by the algorithm. In this implementation, the fitness function must be between 0 and 1, where 1 represents a perfect match to the target and 0 is the furthest distance from the target. In order to generate new solutions, genes are recombined and mutated. A recombination event is when a piece of two genes are swapped between them. A mutations is when a single allele in a gene is changed at random. With encoding/decoding, fitness and evolution pressure, we can produce generations of solutions that tend to include more fit genes within them. The algorithm begins by encoding the input string as a gene. Next, a population of solutions is generated at random. Then we create successive generations of solutions by selecting pairs of parents using tournament selection, randomly applying recombination or mutation to them and then adding them to the next generation. This repeats until a threshold number of generations is reached or the highest fit solution in the generation matches the target. In either case the algorithm returns a dictionary containing the genotype, phenotype and fitness of the solution.* **Uses** [encode_chrarr](#encode_chrarr), [generate_population](#generate_population), [tournament](#tournament), [describe_chromosome](#describe_chromosome), [breed](#breed), and [get_solution](#get_solution)\n",
    "\n",
    "* **population** List[List[str]] - list of chromosomes of potential solution \n",
    "* **target** List[str] - genotype of target\n",
    "* **fitness** Callable - fitness function\n",
    "\n",
    "**returns** Dict - dictionary containing the genotype, phenotype and fitness of a chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1315,
   "metadata": {},
   "outputs": [],
   "source": [
    "def genetic_algorithm(target_str: str, fitness: Callable, parameters: Dict, log_frequency: int = 10) -> Dict:\n",
    "    population_size, tournament_size, max_generation = parameters[\"population_size\"], parameters[\"tournament_size\"], parameters[\"max_generation\"]\n",
    "    crossover_rate, mutate_rate, alphabet = parameters[\"crossover_rate\"], parameters[\"mutate_rate\"], parameters[\"alphabet\"]\n",
    "\n",
    "    target: List[str] = encode_chrarr(target_str)  # encode\n",
    "    population: List[List[str]] = generate_population(len(target), alphabet, population_size)\n",
    "\n",
    "    for generation in range(max_generation):\n",
    "        next_population = []\n",
    "        \n",
    "        solution = get_solution(population, target, fitness)\n",
    "        if solution[\"fitness\"] == 1.0: # stop if we have a match\n",
    "            return solution\n",
    "        \n",
    "        for _ in range(population_size // 2): # fill next generation\n",
    "            parent1, parent2 = tournament(population, target, tournament_size, fitness)\n",
    "            child1, child2 = breed(parent1, parent2, crossover_rate, mutate_rate) # mutate + recombine\n",
    "            next_population.extend([child1, child2])\n",
    "\n",
    "        population = next_population\n",
    "\n",
    "        if generation % log_frequency == 0:\n",
    "            print(f\"{generation:5} {get_solution(population, target, fitness)}\")\n",
    "\n",
    "    return get_solution(population, target, fitness)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1\n",
    "\n",
    "The target is the string \"this is so much fun\".\n",
    "The challenge, aside from implementing the basic algorithm, is deriving a fitness function based on \"b\" - \"p\" (for example).\n",
    "The fitness function should come up with a fitness score based on element to element comparisons between target v. phenotype."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1316,
   "metadata": {},
   "outputs": [],
   "source": [
    "target1 = \"this is so much fun\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1317,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up if you need it.\n",
    "params = {\"population_size\": 100, \"tournament_size\": 10, \"max_generation\": 1000, \"crossover_rate\": 0.9, \"mutate_rate\": 0.05, \"alphabet\": ALPHABET}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1318,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    0 {'genotype': ['g', 'd', 'k', 't', 'r', 'j', 'y', 'u', 'm', 't', ' ', 'o', 'z', 'f', 'q', ' ', 'i', 'q', 'u'], 'fitness': 0.0041841004184100415, 'phenotype': 'gdktrjyumt ozfq iqu'}\n",
      "   10 {'genotype': ['n', 'p', 'k', 'r', ' ', 'a', 't', 'e', 'p', 'h', ' ', 'o', 'r', 'c', 'o', ' ', 'i', 'q', 'r'], 'fitness': 0.007751937984496124, 'phenotype': 'npkr ateph orco iqr'}\n",
      "   20 {'genotype': ['n', 'g', 'k', 'r', ' ', 'a', 't', 'e', 'p', 'h', ' ', 'o', 'r', 'c', 'o', ' ', 'i', 'v', 'q'], 'fitness': 0.00847457627118644, 'phenotype': 'ngkr ateph orco ivq'}\n",
      "   30 {'genotype': ['t', 'g', 'k', 'r', ' ', 'a', 't', 'e', 'r', 'n', ' ', 'o', 'u', 'c', 'o', ' ', 'h', 'u', 'q'], 'fitness': 0.010101010101010102, 'phenotype': 'tgkr atern ouco huq'}\n",
      "   40 {'genotype': ['t', 'g', 'i', 'r', ' ', 'k', 't', 'e', 'r', 'n', ' ', 'o', 'u', 'c', 'd', ' ', 'h', 'u', 'o'], 'fitness': 0.011627906976744186, 'phenotype': 'tgir ktern oucd huo'}\n",
      "   50 {'genotype': ['t', 'g', 'i', 'r', ' ', 'h', 't', ' ', 'r', 'n', ' ', 'o', 'u', 'c', 'j', ' ', 'h', 'u', 'n'], 'fitness': 0.07692307692307693, 'phenotype': 'tgir ht rn oucj hun'}\n",
      "   60 {'genotype': ['t', 'g', 'i', 'r', ' ', 'h', 't', ' ', 'r', 'n', ' ', 'm', 'u', 'c', 'g', ' ', 'h', 'u', 'n'], 'fitness': 0.1, 'phenotype': 'tgir ht rn mucg hun'}\n",
      "   70 {'genotype': ['t', 'g', 'i', 'r', ' ', 'h', 't', ' ', 'r', 'n', ' ', 'm', 'u', 'c', 'h', ' ', 'h', 'u', 'n'], 'fitness': 0.1111111111111111, 'phenotype': 'tgir ht rn much hun'}\n",
      "   80 {'genotype': ['t', 'g', 'i', 'r', ' ', 'h', 't', ' ', 'r', 'n', ' ', 'm', 'u', 'c', 'h', ' ', 'h', 'u', 'n'], 'fitness': 0.1111111111111111, 'phenotype': 'tgir ht rn much hun'}\n",
      "   90 {'genotype': ['t', 'g', 'i', 'r', ' ', 'h', 't', ' ', 'r', 'n', ' ', 'm', 'u', 'c', 'h', ' ', 'h', 'u', 'n'], 'fitness': 0.1111111111111111, 'phenotype': 'tgir ht rn much hun'}\n",
      "  100 {'genotype': ['t', 'h', 'i', 'r', ' ', 'h', 't', ' ', 'r', 'n', ' ', 'm', 'u', 'c', 'h', ' ', 'h', 'u', 'n'], 'fitness': 0.125, 'phenotype': 'thir ht rn much hun'}\n",
      "  110 {'genotype': ['t', 'h', 'i', 'r', ' ', 'h', 't', ' ', 's', 'n', ' ', 'm', 'u', 'c', 'h', ' ', 'h', 'u', 'n'], 'fitness': 0.14285714285714285, 'phenotype': 'thir ht sn much hun'}\n",
      "  120 {'genotype': ['t', 'h', 'i', 'r', ' ', 'h', 't', ' ', 's', 'n', ' ', 'm', 'u', 'c', 'h', ' ', 'h', 'u', 'n'], 'fitness': 0.14285714285714285, 'phenotype': 'thir ht sn much hun'}\n",
      "  130 {'genotype': ['t', 'h', 'i', 'r', ' ', 'h', 't', ' ', 's', 'n', ' ', 'm', 'u', 'c', 'h', ' ', 'e', 'u', 'n'], 'fitness': 0.16666666666666666, 'phenotype': 'thir ht sn much eun'}\n",
      "  140 {'genotype': ['t', 'h', 'i', 'r', ' ', 'h', 't', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'e', 'u', 'n'], 'fitness': 0.2, 'phenotype': 'thir ht so much eun'}\n",
      "  150 {'genotype': ['t', 'h', 'i', 'r', ' ', 'h', 't', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'e', 'u', 'n'], 'fitness': 0.2, 'phenotype': 'thir ht so much eun'}\n",
      "  160 {'genotype': ['t', 'h', 'i', 'r', ' ', 'i', 't', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'e', 'u', 'n'], 'fitness': 0.25, 'phenotype': 'thir it so much eun'}\n",
      "  170 {'genotype': ['t', 'h', 'i', 'r', ' ', 'i', 't', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'e', 'u', 'n'], 'fitness': 0.25, 'phenotype': 'thir it so much eun'}\n",
      "  180 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 't', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'e', 'u', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'this it so much eun'}\n",
      "  190 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 't', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'e', 'u', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'this it so much eun'}\n",
      "  200 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 't', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'e', 'u', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'this it so much eun'}\n",
      "  210 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 't', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'e', 'u', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'this it so much eun'}\n",
      "  220 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 't', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'e', 'u', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'this it so much eun'}\n",
      "  230 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 't', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'e', 'u', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'this it so much eun'}\n",
      "  240 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 't', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'e', 'u', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'this it so much eun'}\n",
      "  250 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 't', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'e', 'u', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'this it so much eun'}\n",
      "  260 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 't', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'e', 'u', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'this it so much eun'}\n",
      "  270 {'genotype': ['t', 'h', 'i', 's', ' ', 'i', 't', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 0.5, 'phenotype': 'this it so much fun'}\n"
     ]
    }
   ],
   "source": [
    "result1 = genetic_algorithm(target1, fitness_forward, params)  # do what you need to do for your implementation but don't change the lines above or below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1319,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'fitness': 1.0,\n",
      " 'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u',\n",
      "              'c', 'h', ' ', 'f', 'u', 'n'],\n",
      " 'phenotype': 'this is so much fun'}\n"
     ]
    }
   ],
   "source": [
    "pprint(result1, compact=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2\n",
    "\n",
    "You should have working code now.\n",
    "The goal here is to think a bit more about fitness functions.\n",
    "The target string is now, 'nuf hcum os si siht'.\n",
    "This is obviously target #1 but reversed.\n",
    "If we just wanted to match the string, this would be trivial.\n",
    "Instead, this problem, we want to \"decode\" the string so that the best individual displays the target forwards.\n",
    "In order to do this, you'll need to come up with a fitness function that measures how successful candidates are towards this goal.\n",
    "The constraint is that you may not perform any global operations on the target or individuals.\n",
    "Your fitness function must still compare a single gene against a single gene.\n",
    "Your solution will likely not be Pythonic but use indexing.\n",
    "That's ok.\n",
    "<div style=\"background: lemonchiffon; margin:20px; padding: 20px;\">\n",
    "    <strong>Important</strong>\n",
    "    <p>\n",
    "        You may not reverse an entire string (either target or candidate) at any time.\n",
    "        Everything must be a computation of one gene against one gene (one letter against one letter).\n",
    "        Failure to follow these directions will result in 0 points for the problem.\n",
    "    </p>\n",
    "</div>\n",
    "\n",
    "The best individual in the population is the one who expresses this string *forwards*.\n",
    "\n",
    "\"this is so much fun\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1320,
   "metadata": {},
   "outputs": [],
   "source": [
    "target2 = \"nuf hcum os si siht\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1321,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up if you need it.\n",
    "params = {\"population_size\": 100, \"tournament_size\": 10, \"max_generation\": 1000, \"crossover_rate\": 0.9, \"mutate_rate\": 0.05, \"alphabet\": ALPHABET}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1322,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    0 {'genotype': ['m', 'x', 'i', 'u', ' ', 'w', 'y', 'd', 'q', 'p', 'h', 'y', 'w', 'f', 'v', 'g', 'g', 'q', 'm'], 'fitness': 0.003367003367003367, 'phenotype': 'mxiu wydqphywfvggqm'}\n",
      "   10 {'genotype': ['z', 'e', 'j', 'u', ' ', 'c', 's', ' ', 'q', 'r', ' ', 'k', 'v', 'g', 'g', ' ', 'e', 'q', 'm'], 'fitness': 0.02631578947368421, 'phenotype': 'zeju cs qr kvgg eqm'}\n",
      "   20 {'genotype': ['v', 'e', 'j', 'u', ' ', 'f', 's', ' ', 'q', 'o', ' ', 'k', 'v', 'g', 'g', ' ', 'e', 'v', 'm'], 'fitness': 0.04, 'phenotype': 'veju fs qo kvgg evm'}\n",
      "   30 {'genotype': ['v', 'j', 'j', 's', ' ', 'f', 's', ' ', 'q', 'o', ' ', 'k', 'u', 'e', 'g', ' ', 'e', 'v', 'm'], 'fitness': 0.05263157894736842, 'phenotype': 'vjjs fs qo kueg evm'}\n",
      "   40 {'genotype': ['v', 'j', 'j', 's', ' ', 'f', 's', ' ', 's', 'o', ' ', 'k', 'u', 'e', 'g', ' ', 'e', 'v', 'm'], 'fitness': 0.058823529411764705, 'phenotype': 'vjjs fs so kueg evm'}\n",
      "   50 {'genotype': ['v', 'j', 'j', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'k', 'u', 'e', 'g', ' ', 'e', 'v', 'm'], 'fitness': 0.06666666666666667, 'phenotype': 'vjjs js so kueg evm'}\n",
      "   60 {'genotype': ['t', 'j', 'j', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'k', 'u', 'e', 'g', ' ', 'e', 'v', 'm'], 'fitness': 0.07692307692307693, 'phenotype': 'tjjs js so kueg evm'}\n",
      "   70 {'genotype': ['t', 'h', 'j', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'k', 'u', 'e', 'g', ' ', 'e', 'v', 'm'], 'fitness': 0.09090909090909091, 'phenotype': 'thjs js so kueg evm'}\n",
      "   80 {'genotype': ['t', 'h', 'j', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'k', 'u', 'e', 'g', ' ', 'e', 'v', 'n'], 'fitness': 0.1, 'phenotype': 'thjs js so kueg evn'}\n",
      "   90 {'genotype': ['t', 'h', 'j', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'k', 'u', 'e', 'g', ' ', 'e', 'v', 'n'], 'fitness': 0.1, 'phenotype': 'thjs js so kueg evn'}\n",
      "  100 {'genotype': ['t', 'h', 'j', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'k', 'u', 'e', 'g', ' ', 'e', 'v', 'n'], 'fitness': 0.1, 'phenotype': 'thjs js so kueg evn'}\n",
      "  110 {'genotype': ['t', 'h', 'j', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'k', 'u', 'e', 'g', ' ', 'f', 'v', 'n'], 'fitness': 0.1111111111111111, 'phenotype': 'thjs js so kueg fvn'}\n",
      "  120 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'k', 'u', 'e', 'g', ' ', 'f', 'u', 'n'], 'fitness': 0.14285714285714285, 'phenotype': 'this js so kueg fun'}\n",
      "  130 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'k', 'u', 'e', 'g', ' ', 'f', 'u', 'n'], 'fitness': 0.14285714285714285, 'phenotype': 'this js so kueg fun'}\n",
      "  140 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'k', 'u', 'c', 'g', ' ', 'f', 'u', 'n'], 'fitness': 0.2, 'phenotype': 'this js so kucg fun'}\n",
      "  150 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'l', 'u', 'c', 'g', ' ', 'f', 'u', 'n'], 'fitness': 0.25, 'phenotype': 'this js so lucg fun'}\n",
      "  160 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'l', 'u', 'c', 'g', ' ', 'f', 'u', 'n'], 'fitness': 0.25, 'phenotype': 'this js so lucg fun'}\n",
      "  170 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'l', 'u', 'c', 'g', ' ', 'f', 'u', 'n'], 'fitness': 0.25, 'phenotype': 'this js so lucg fun'}\n",
      "  180 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'l', 'u', 'c', 'g', ' ', 'f', 'u', 'n'], 'fitness': 0.25, 'phenotype': 'this js so lucg fun'}\n",
      "  190 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'l', 'u', 'c', 'g', ' ', 'f', 'u', 'n'], 'fitness': 0.25, 'phenotype': 'this js so lucg fun'}\n",
      "  200 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'l', 'u', 'c', 'g', ' ', 'f', 'u', 'n'], 'fitness': 0.25, 'phenotype': 'this js so lucg fun'}\n",
      "  210 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'g', ' ', 'f', 'u', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'this js so mucg fun'}\n",
      "  220 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'g', ' ', 'f', 'u', 'n'], 'fitness': 0.3333333333333333, 'phenotype': 'this js so mucg fun'}\n",
      "  230 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 0.5, 'phenotype': 'this js so much fun'}\n",
      "  240 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 0.5, 'phenotype': 'this js so much fun'}\n",
      "  250 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 0.5, 'phenotype': 'this js so much fun'}\n",
      "  260 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 0.5, 'phenotype': 'this js so much fun'}\n",
      "  270 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 0.5, 'phenotype': 'this js so much fun'}\n",
      "  280 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 0.5, 'phenotype': 'this js so much fun'}\n",
      "  290 {'genotype': ['t', 'h', 'i', 's', ' ', 'j', 's', ' ', 's', 'o', ' ', 'm', 'u', 'c', 'h', ' ', 'f', 'u', 'n'], 'fitness': 0.5, 'phenotype': 'this js so much fun'}\n"
     ]
    }
   ],
   "source": [
    "result2 = genetic_algorithm(target2, fitness_reverse, params)  # do what you need to do for your implementation but don't change the lines above or below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1323,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'fitness': 1.0,\n",
      " 'genotype': ['t', 'h', 'i', 's', ' ', 'i', 's', ' ', 's', 'o', ' ', 'm', 'u',\n",
      "              'c', 'h', ' ', 'f', 'u', 'n'],\n",
      " 'phenotype': 'this is so much fun'}\n"
     ]
    }
   ],
   "source": [
    "pprint(result2, compact=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3\n",
    "\n",
    "This is a variation on the theme of Problem 2.\n",
    "The Caeser Cypher replaces each letter of a string with the letter 13 characters down alphabet (rotating from \"z\" back to \"a\" as needed).\n",
    "This is also known as ROT13 (for \"rotate 13\").\n",
    "Latin did not have spaces (and the space is not continguous with the letters a-z) so we'll remove them from our alphabet.\n",
    "Again, the goal is to derive a fitness function that compares a single gene against a single gene, without global transformations.\n",
    "This fitness function assigns higher scores to individuals that correctly decode the target.\n",
    "\n",
    "<div style=\"background: lemonchiffon; margin:20px; padding: 20px;\">\n",
    "    <strong>Important</strong>\n",
    "    <p>\n",
    "        You may not apply ROT13 to an entire string (either target or candidate) at any time.\n",
    "        Everything must be a computation of one gene against one gene.\n",
    "        Failure to follow these directions will result in 0 points for the problem.\n",
    "    </p>\n",
    "</div>\n",
    "\n",
    "The best individual will express the target *decoded*.\n",
    "\n",
    "\"thisissomuchfun\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1324,
   "metadata": {},
   "outputs": [],
   "source": [
    "ALPHABET3 = \"abcdefghijklmnopqrstuvwxyz\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1325,
   "metadata": {},
   "outputs": [],
   "source": [
    "target3 = \"guvfvffbzhpusha\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1326,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set up if you need it\n",
    "params = {\"population_size\": 100, \"tournament_size\": 10, \"max_generation\": 1000, \"crossover_rate\": 0.9, \"mutate_rate\": 0.05, \"alphabet\": ALPHABET3}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1327,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    0 {'genotype': ['q', 'k', 'h', 'm', 'h', 'w', 'w', 'q', 'b', 'u', 'd', 'j', 'c', 'v', 'v'], 'fitness': 0.015384615384615385, 'phenotype': 'qkhmhwwqbudjcvv'}\n",
      "   10 {'genotype': ['t', 'i', 'i', 'o', 'i', 'u', 'p', 'q', 'l', 'u', 'd', 'j', 'd', 'v', 'q'], 'fitness': 0.043478260869565216, 'phenotype': 'tiioiupqludjdvq'}\n",
      "   20 {'genotype': ['t', 'i', 'i', 's', 'i', 'u', 'p', 'q', 'l', 'u', 'd', 'g', 'f', 'v', 'o'], 'fitness': 0.07142857142857142, 'phenotype': 'tiisiupqludgfvo'}\n",
      "   30 {'genotype': ['t', 'i', 'i', 's', 'i', 'r', 'u', 'p', 'l', 'u', 'd', 'g', 'f', 'v', ' '], 'fitness': 0.1, 'phenotype': 'tiisirupludgfv '}\n",
      "   40 {'genotype': ['t', 'i', 'i', 's', 'i', 'r', 'r', 'p', 'l', 'u', 'd', 'g', 'f', 'v', ' '], 'fitness': 0.1111111111111111, 'phenotype': 'tiisirrpludgfv '}\n",
      "   50 {'genotype': ['t', 'i', 'i', 's', 'i', 'r', 'r', 'p', 'l', 'u', 'd', 'g', 'f', 'v', ' '], 'fitness': 0.1111111111111111, 'phenotype': 'tiisirrpludgfv '}\n",
      "   60 {'genotype': ['t', 'i', 'i', 's', 'i', 'r', 'r', 'p', 'l', 'u', 'd', 'g', 'f', 'v', ' '], 'fitness': 0.1111111111111111, 'phenotype': 'tiisirrpludgfv '}\n",
      "   70 {'genotype': ['t', 'h', 'i', 's', 'i', 'r', 's', 'p', 'l', 'u', 'd', 'g', 'f', 'v', ' '], 'fitness': 0.14285714285714285, 'phenotype': 'thisirspludgfv '}\n",
      "   80 {'genotype': ['t', 'h', 'i', 's', 'i', 'r', 's', 'p', 'l', 'u', 'd', 'g', 'f', 'v', ' '], 'fitness': 0.14285714285714285, 'phenotype': 'thisirspludgfv '}\n",
      "   90 {'genotype': ['t', 'h', 'i', 's', 'i', 'r', 's', 'p', 'l', 'u', 'd', 'h', 'f', 'v', ' '], 'fitness': 0.16666666666666666, 'phenotype': 'thisirspludhfv '}\n",
      "  100 {'genotype': ['t', 'h', 'i', 's', 'i', 'r', 's', 'p', 'l', 'u', 'd', 'h', 'f', 'v', ' '], 'fitness': 0.16666666666666666, 'phenotype': 'thisirspludhfv '}\n",
      "  110 {'genotype': ['t', 'h', 'i', 's', 'i', 'r', 's', 'p', 'l', 'u', 'd', 'h', 'f', 'v', ' '], 'fitness': 0.16666666666666666, 'phenotype': 'thisirspludhfv '}\n",
      "  120 {'genotype': ['t', 'h', 'i', 's', 'i', 'r', 's', 'p', 'l', 'u', 'd', 'h', 'f', 'v', ' '], 'fitness': 0.16666666666666666, 'phenotype': 'thisirspludhfv '}\n",
      "  130 {'genotype': ['t', 'h', 'i', 's', 'i', 'r', 's', 'p', 'l', 'u', 'd', 'h', 'f', 'v', ' '], 'fitness': 0.16666666666666666, 'phenotype': 'thisirspludhfv '}\n",
      "  140 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'l', 'u', 'd', 'h', 'f', 'v', ' '], 'fitness': 0.25, 'phenotype': 'thisissoludhfv '}\n",
      "  150 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'l', 'u', 'd', 'h', 'f', 'v', ' '], 'fitness': 0.25, 'phenotype': 'thisissoludhfv '}\n",
      "  160 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'l', 'u', 'd', 'h', 'f', 'u', ' '], 'fitness': 0.3333333333333333, 'phenotype': 'thisissoludhfu '}\n",
      "  170 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'l', 'u', 'd', 'h', 'f', 'u', ' '], 'fitness': 0.3333333333333333, 'phenotype': 'thisissoludhfu '}\n",
      "  180 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'l', 'u', 'd', 'h', 'f', 'u', ' '], 'fitness': 0.3333333333333333, 'phenotype': 'thisissoludhfu '}\n",
      "  190 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'l', 'u', 'd', 'h', 'f', 'u', ' '], 'fitness': 0.3333333333333333, 'phenotype': 'thisissoludhfu '}\n",
      "  200 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'l', 'u', 'd', 'h', 'f', 'u', ' '], 'fitness': 0.3333333333333333, 'phenotype': 'thisissoludhfu '}\n",
      "  210 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'd', 'h', 'f', 'u', ' '], 'fitness': 0.5, 'phenotype': 'thisissomudhfu '}\n",
      "  220 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'd', 'h', 'f', 'u', ' '], 'fitness': 0.5, 'phenotype': 'thisissomudhfu '}\n",
      "  230 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'd', 'h', 'f', 'u', ' '], 'fitness': 0.5, 'phenotype': 'thisissomudhfu '}\n",
      "  240 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'd', 'h', 'f', 'u', ' '], 'fitness': 0.5, 'phenotype': 'thisissomudhfu '}\n",
      "  250 {'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'd', 'h', 'f', 'u', ' '], 'fitness': 0.5, 'phenotype': 'thisissomudhfu '}\n"
     ]
    }
   ],
   "source": [
    "result3 = genetic_algorithm(target3, fitness_caeser, params)  # do what you need to do for your implementation but don't change the lines above or below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1328,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'fitness': 1.0,\n",
      " 'genotype': ['t', 'h', 'i', 's', 'i', 's', 's', 'o', 'm', 'u', 'c', 'h', 'f',\n",
      "              'u', ' '],\n",
      " 'phenotype': 'thisissomuchfu '}\n"
     ]
    }
   ],
   "source": [
    "pprint(result3, compact=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4\n",
    "\n",
    "There is no code for this problem.\n",
    "\n",
    "In Problem 3, we assumed we knew what the shift was in ROT-13.\n",
    "What if we didn't?\n",
    "Describe how you might solve that problem including a description of the solution encoding (chromosome and interpretation) and fitness function. Assume we can add spaces into the message."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**answer here**\n",
    "\n",
    "In order to capture the transformation needed to translate between the genotype and target, we can add another gene (or set of genes) to store transformation information. For string matching, we could store an extra gene that encode the amount to rotate the string. We would then adjust the fitness function to treat the first n genes as normal characters and the last one as a rotation amount. We can then use the machinery of the genetic algorithm to drive towards solutions that have the correct transformation. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Challenge\n",
    "\n",
    "**You do not need to do this problem and it won't be graded if you do. It's just here if you want to push your understanding.**\n",
    "\n",
    "The original GA used binary encodings for everything.\n",
    "We're basically using a Base 27 encoding.\n",
    "You could, however, write a version of the algorithm that uses an 8 bit encoding for each letter (ignore spaces as they're a bit of a bother).\n",
    "That is, a 4 letter candidate looks like this:\n",
    "\n",
    "```\n",
    "0, 1, 1, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1\n",
    "```\n",
    "\n",
    "If you wrote your `genetic_algorithm` code general enough, with higher order functions, you should be able to implement it using bit strings instead of latin strings."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Before You Submit...\n",
    "\n",
    "1. Did you provide output exactly as requested?\n",
    "2. Did you re-execute the entire notebook? (\"Restart Kernel and Rull All Cells...\")\n",
    "3. If you did not complete the assignment or had difficulty please explain what gave you the most difficulty in the Markdown cell below.\n",
    "4. Did you change the name of the file to `jhed_id.ipynb`?\n",
    "\n",
    "Do not submit any other files."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "en605645",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.14"
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "120px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": false,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
